File: packages/parser/src/crypto.ts
/**
 * Crypto layer for ACLOS WonderfulDb.
 *
 * File format (verified against ACLOS 2.15.3.449, @tencent/icreate-wonderful@3.5.2):
 *   - file body is hex-encoded ASCII (0-9, a-f)
 *   - decoded bytes = AES-256-CBC ciphertext (PKCS7 padding, stripped by Web Crypto)
 *   - key = sha256(openid).hex.substring(0, 32) as UTF-8 bytes  (32 bytes)
 *   - iv  = sha256(openid).hex.substring(0, 16) as UTF-8 bytes  (16 bytes)
 *   - plaintext = JSON object: { "key_wonderful_list_<openid>": Match[] }
 *
 * Uses Web Crypto (globalThis.crypto.subtle), so it works in Node 20+, Bun,
 * and WebView2 with no Node-specific imports. Async because subtle is async.
 *
 * Reference: packages/parser/src/schema/_acl-source/safeUtils.js
 */

const KEY_HEX_LEN = 32;  // sha256 hex prefix for AES-256 key
const IV_HEX_LEN = 16;   // sha256 hex prefix for IV

async function sha256Hex(s: string): Promise<string> {
  const data = new TextEncoder().encode(s);
  const hashBuf = await crypto.subtle.digest('SHA-256', data);
  const bytes = new Uint8Array(hashBuf);
  let hex = '';
  for (let i = 0; i < bytes.length; i++) {
    const b = bytes[i]!;
    hex += (b >>> 4).toString(16) + (b & 0x0f).toString(16);
  }
  return hex;
}

export async function deriveKeyIv(openid: string): Promise<{ key: Uint8Array; iv: Uint8Array }> {
  const sha = await sha256Hex(openid);
  return {
    key: new TextEncoder().encode(sha.substring(0, KEY_HEX_LEN)),
    iv: new TextEncoder().encode(sha.substring(0, IV_HEX_LEN)),
  };
}

export async function aesDecrypt(ciphertext: Uint8Array, key: Uint8Array, iv: Uint8Array): Promise<Uint8Array> {
  const cryptoKey = await crypto.subtle.importKey(
    'raw',
    key as BufferSource,
    { name: 'AES-CBC' },
    false,
    ['decrypt'],
  );
  const plainBuf = await crypto.subtle.decrypt(
    { name: 'AES-CBC', iv: iv as BufferSource },
    cryptoKey,
    ciphertext as BufferSource,
  );
  return new Uint8Array(plainBuf);
}

/** Decrypt a WonderfulDb buffer (= decoded hex bytes) given the openid. Returns the JSON plaintext. */
export async function decryptWonderfulDbBuffer(buf: Uint8Array, openid: string): Promise<string> {
  const { key, iv } = await deriveKeyIv(openid);
  const plain = await aesDecrypt(buf, key, iv);
  return new TextDecoder().decode(plain);
}

File: src-tauri/src/parser/crypto.rs
//! Crypto layer for ACLOS WonderfulDb.
//!
//! File format (verified against ACLOS 2.15.3.449, @tencent/icreate-wonderful@3.5.2):
//!   - file body is hex-encoded ASCII (0-9, a-f)
//!   - decoded bytes = AES-256-CBC ciphertext (PKCS7 padding, stripped by cbc crate)
//!   - key = sha256(openid).hex.substring(0, 32) as UTF-8 bytes  (32 bytes)
//!   - iv  = sha256(openid).hex.substring(0, 16) as UTF-8 bytes  (16 bytes)
//!   - plaintext = JSON object: { "key_wonderful_list_<openid>": Match[] }

use crate::parser::error::ParseError;
use aes::Aes256;
use cbc::cipher::{block_padding::Pkcs7, BlockDecryptMut, KeyIvInit};
use sha2::{Digest, Sha256};

type Aes256CbcDec = cbc::Decryptor<Aes256>;

/// Derive the AES-256 key (32 bytes) and CBC IV (16 bytes) from the openid.
/// Both are the first 32 / 16 hex chars of sha256(openid), interpreted as
/// UTF-8 bytes — i.e. literal ASCII hex chars, not the decoded bytes.
pub fn derive_key_iv(openid: &str) -> (Vec<u8>, Vec<u8>) {
    let mut hasher = Sha256::new();
    hasher.update(openid.as_bytes());
    let sha_hex = hex::encode(hasher.finalize());
    let key = sha_hex.as_bytes()[..32].to_vec();
    let iv = sha_hex.as_bytes()[..16].to_vec();
    (key, iv)
}

/// Decrypt AES-256-CBC ciphertext with PKCS7 padding. Returns the plaintext.
pub fn aes_decrypt(cipher: &[u8], key: &[u8], iv: &[u8]) -> Result<Vec<u8>, ParseError> {
    let decryptor = Aes256CbcDec::new_from_slices(key, iv)
        .map_err(|e| ParseError::Crypto(format!("invalid key/iv: {}", e)))?;
    let mut buf = cipher.to_vec();
    let pt = decryptor
        .decrypt_padded_mut::<Pkcs7>(&mut buf)
        .map_err(|e| ParseError::Crypto(format!("decrypt failed: {}", e)))?;
    Ok(pt.to_vec())
}
